### PR TITLE
Create a Triangulation<3> from a Triangulation<2,3>

### PR DESCRIPTION
**Pull request by fdrmrc**
This creates a Triangulation<3> out of a Triangulation<2,3>, filling the latter with tetrahedrons. Named parameters can be used to control the size and the quality of the final mesh. 
I put this into `GridTools`, but  I'm happy to move this somewhere else if you think it's not the right namespace/place.

The .cc test is just filling an hypersphere, but one could use this after boolean operations (#13717) , giving the capability of meshing things like the following, where the union of two spheres is computed and meshed inside
boolean_spheres_union

### CODE REVIEW

**Comment by luca-heltai** [comment]
I think this should go in `grid_geneerator.h`, next to its "inverse", i.e., the function that extracts a boundary mesh from a Triangulation.

File: include/deal.II/grid/grid_tools.h [hunk 1]
```diff
@@ -576,6 +581,28 @@ namespace GridTools
   transform(const Transformation &        transformation,
             Triangulation<dim, spacedim> &triangulation);
 
+#ifdef DEAL_II_WITH_CGAL
+  /**
+   * Create a deal.II Triangulation<3> out of a deal.II Triangulation<2,3>
+   * by filling it with tetrahedrons. Optional variable arguments `cgal_args`
+   * can be used to control size, quality, and distribution of the cells of
+   * the final triangulation.
```
Comment (luca-heltai) [inline include/deal.II/grid/grid_tools.h:589]: Can you copy here the blob about the parameters of mesh_criteria_3 that we have in the other function?

File: tests/cgal/cgal_triangulation_07.cc [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    CGAL::parameters::facet_size             = 0.1,
+    CGAL::parameters::facet_distance         = 0,
+    CGAL::parameters::cell_radius_edge_ratio = 2,
+    CGAL::parameters::cell_size              = 0.1);
+
+  GridOut       go;
+  std::ofstream out_tria_name("tria_surface_to_volumetric.vtk");
+  go.write_vtk(out_tria, out_tria_name);
+
+  // cat_file("tria_surface_to_volumetric.vtk");
```
Comment (luca-heltai) [inline tests/cgal/cgal_triangulation_07.cc:53]:
```suggestion
  remove("tria_surface_to_volumetric.vtk");
```

**Review by luca-heltai — CHANGES_REQUESTED** [review]
Besides moving it to `grid_generator.h`, I think the documentation should be filled with the information about the additional parameters.

File: include/deal.II/grid/grid_tools.h [hunk 3]
```diff
@@ -576,6 +581,28 @@ namespace GridTools
...
+   * @param [in] surface_tria The input deal.II Triangulation<2,3>.
+   * @param [out] vol_tria The output deal.II Triangulation<3>.
+   * @param [in] cgal_args Additional parameters to pass to the CGAL::make_mesh_3
+   * function.
+   */
+  template <typename... Args>
+  void
+  surface_mesh_to_volumetric_mesh(const Triangulation<2, 3> &surface_tria,
+                                  Triangulation<3> &         vol_tria,
+                                  Args... cgal_args);
```
Comment (peterrum) [inline include/deal.II/grid/grid_tools.h]: Personally, I would prefer to introduce an `AdditionalData`.

File: include/deal.II/grid/grid_tools.h [hunk 4]
```diff
@@ -911,6 +938,13 @@ namespace GridTools
   regularize_corner_cells(Triangulation<dim, spacedim> &tria,
                           const double limit_angle_fraction = .75);
 
+#ifdef DEAL_II_WITH_CGAL
+  template <typename... Args>
+  void
+  surface_mesh_to_volumetric_mesh(const Triangulation<2, 3> &surface_tria,
+                                  Triangulation<3> &         vol_tria,
+                                  Args... cgal_args);
+#endif
```
Comment (peterrum) [inline include/deal.II/grid/grid_tools.h]: Why is this here? Isn't this the same as above?

File: include/deal.II/grid/grid_tools.h (hunk 3 above)
Comment (peterrum) [reply include/deal.II/grid/grid_tools.h]: Can we move `#ifdef DEAL_II_WITH_CGAL` within the function and assert it there?

**Review by peterrum — COMMENTED** [review]
Very nice!

**Comment by fdrmrc** [comment]
@peterrum I pushed right after you wrote the review!

**Comment by peterrum** [comment]
One question: the original surface mesh determines the cells of the volume mesh on the surface or are the cells unrelated?

**Comment by fdrmrc** [comment]
> One question: the original surface mesh determines the cells of the volume mesh on the surface or are the cells unrelated?

They're currently unrelated. The use case I had in mind is like the one in the test: one just starts with a deal.II tria describing the surface and fills it. It should not be a problem however to record the topological information about the first deal.II surface tria and use it to create the volumetric one.

File: include/deal.II/grid/grid_generator.h [hunk 5]
```diff
@@ -1746,6 +1758,40 @@ namespace GridGenerator
...
+   * @param [in] surface_tria The input deal.II Triangulation<2,3>.
+   * @param [out] vol_tria The output deal.II Triangulation<3>.
+   * @param [in] cgal_args Additional parameters to pass to the CGAL::make_mesh_3
+   * function.
+   */
+  template <typename... Args>
+  void
+  surface_mesh_to_volumetric_mesh(const Triangulation<2, 3> &surface_tria,
+                                  Triangulation<3> &         vol_tria,
+                                  Args... cgal_args);
```
[the remark this answers is not in the document]
Comment (fdrmrc) [reply include/deal.II/grid/grid_generator.h:1793]: Sure. As soon as the other patch is okay, I can update this!

**Comment by fdrmrc** [comment]
So done, this should be good to go now.  
Since the function is not templated anymore, I've moved it inside `grid_generator.cc`. 

Also, I took the liberty to move the two utility functions that convert points back and forth from deal.II to CGAL in a separate header, fixing some issues with headers inclusions I hadn't noticed before.

File: source/grid/grid_generator.cc [hunk 6]
```diff
@@ -8639,6 +8639,63 @@ namespace GridGenerator
...
+                           CGAL::parameters::facet_distance =
+                             data.facet_distance,
+                           CGAL::parameters::cell_radius_edge_ratio =
+                             data.cell_radius_edge_ratio,
+                           CGAL::parameters::cell_size = data.cell_size);
+    const auto cgal_triangulation = CGAL::make_mesh_3<C3t3>(domain, criteria);
+    CGALWrappers::cgal_triangulation_to_dealii_triangulation(cgal_triangulation,
+                                                             vol_tria);
+  }
+#  endif
```
Comment (luca-heltai) [inline source/grid/grid_generator.cc:8697]: Can you add a default implementation that throws an error saying that the function needs CGAL to be installed?

**Review by luca-heltai — APPROVED** [review]
Minor change required: add an implementation stub that throws ExcNotImplemented if cgal is not found.

File: include/deal.II/grid/grid_generator.h [hunk 7]
```diff
@@ -1823,6 +1824,38 @@ namespace GridGenerator
                       CGALWrappers::AdditionalData<dim>{},
                     const Point<3> &interior_point    = Point<3>(),
                     const double &  outer_ball_radius = 1.0);
+
+  /**
+   * Create a deal.II Triangulation<3> out of a deal.II Triangulation<2,3>
+   * by filling it with tetrahedrons.
```
Comment (kronbichler) [inline include/deal.II/grid/grid_generator.h:1830]:
```suggestion
   * by filling it with tetrahedra.
```

File: include/deal.II/grid/grid_generator.h [hunk 8]
```diff
@@ -1823,6 +1824,38 @@ namespace GridGenerator
...
+   *
+   * The optional variable arguments @p cgal_args can be used to pass additional
+   * arguments to the CGAL::Mesh_criteria_3 class (see
+   * https://doc.cgal.org/latest/Mesh_3/index.html) for more information.
```
Comment (kronbichler) [inline include/deal.II/grid/grid_generator.h:1834]: Can you please adapt this and the text below, as we have now introduced AdditionalData.

**Review by kronbichler — APPROVED** [review]
Looks good, apart from some adaptations to the comments.

**Comment by fdrmrc** [comment]
I'm addressing Luca's comment, please don't merge

**Comment by fdrmrc** [comment]
I rebased this PR onto the smaller ones #13928  and #13929, so that here there's effectively only the implementation of `surface_tria_to_volumetric()`.

**Comment by fdrmrc** [comment]
Rebased on top of master.

**Review by kronbichler — APPROVED** [review]

**Comment by fdrmrc** [comment]
The implementation of the present function has been moved in `grid_generator_cgal.cc`, following #13954

**Review by kronbichler — APPROVED** [review]

**Pull request merged**